### PR TITLE
Feature/websign

### PR DESCRIPTION
**Pull request by meletios**

### CODE REVIEW

File: member-service/src/main/java/com/hedvig/memberservice/config/QuartzConfig.java [hunk 1]
```diff
@@ -1,38 +1,33 @@
...
 
-  @Bean
-  Trigger echoTrigger(JobDetail echoJob) {
-    int frequencyInSec = 10;
-    log.info("Configuring trigger");
+    try {
 
-    return newTrigger().withIdentity("echoTrigger").forJob(echoJob).withSchedule(simpleSchedule().withIntervalInSeconds(frequencyInSec).repeatForever()).build();
+      if (scheduler.deleteJob(new JobKey("echo3", "test"))) {
+        log.info("Quartz job eco3 was deleted");
```
Comment (meletios) [inline member-service/src/main/java/com/hedvig/memberservice/config/QuartzConfig.java]: 'eco3' should be 'echo3' :wink:

File: member-service/src/main/java/com/hedvig/memberservice/jobs/BankIdCollector.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import org.quartz.JobExecutionContext;
+import org.quartz.JobExecutionException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.scheduling.quartz.QuartzJobBean;
+
+public class BankIdCollector extends QuartzJobBean {
+
+  static final Logger log = LoggerFactory.getLogger(EchoJob.class);
```
Comment (meletios) [inline member-service/src/main/java/com/hedvig/memberservice/jobs/BankIdCollector.java]: Should be BankIdCollector.class

File: member-service/src/main/java/com/hedvig/memberservice/jobs/BankIdCollector.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  @Autowired
+  SigningService signingService;
+
+  @Override
+  protected void executeInternal(JobExecutionContext context) throws JobExecutionException {
+    try {
+      val referenceId = context.getJobDetail().getKey().getName();
+      log.info("Starting BankIdCollectorJob for {}", referenceId);
+
+      val a = signingService.collectBankId(referenceId);
```
Comment (meletios) [inline member-service/src/main/java/com/hedvig/memberservice/jobs/BankIdCollector.java]: rename to much more informative name

File: member-service/src/main/java/com/hedvig/memberservice/query/StaticEventSender.java [hunk 4]
```diff
@@ -17,20 +18,29 @@
 
   private final ProductApi productApi;
   private Logger logger = LoggerFactory.getLogger(StaticEventSender.class);
+  private SigningService signService;
 
   @Autowired
-  public StaticEventSender(MemberRepository memberRepo, ProductApi productApi) {
+  public StaticEventSender(MemberRepository memberRepo, ProductApi productApi,
```
Comment (meletios) [inline member-service/src/main/java/com/hedvig/memberservice/query/StaticEventSender.java]: We dont use the memberRepo

File: member-service/src/main/java/com/hedvig/memberservice/services/BankIdRestService.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package com.hedvig.memberservice.services;
+
+import com.hedvig.external.bankID.bankIdRestTypes.CollectResponse;
+import com.hedvig.external.bankID.bankIdRestTypes.OrderResponse;
+import org.springframework.lang.NonNull;
+
+public interface BankIdRestService {
+
+  OrderResponse startSign(final long memberId, final @NonNull String ssn, final @NonNull String userVisibleMessage, final @NonNull String ipAddress);
```
Comment (meletios) [inline member-service/src/main/java/com/hedvig/memberservice/services/BankIdRestService.java]: we dont user the memberId

File: member-service/src/main/java/com/hedvig/memberservice/services/BankIdRestServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+  private final BankIdRestApi bankIdRestApi;
+
+
+  public BankIdRestServiceImpl(BankIdRestApi bankIdRestApi) {
+    this.bankIdRestApi = bankIdRestApi;
+  }
+
+  @Override
+  public OrderResponse startSign(long memberId, @NonNull final String ssn, @NonNull final String userVisibleMessage, @NonNull final String ipNumber) {
```
Comment (meletios) [inline member-service/src/main/java/com/hedvig/memberservice/services/BankIdRestServiceImpl.java]: memberId is not used

File: member-service/src/main/java/com/hedvig/memberservice/services/SchedulerService.java [hunk 7]
```diff
@@ -1,14 +1,7 @@
 package com.hedvig.memberservice.services;
```
Comment (meletios) [inline member-service/src/main/java/com/hedvig/memberservice/services/SchedulerService.java]: Remove file

File: member-service/src/main/java/com/hedvig/memberservice/services/SigningService.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+  public MemberSignResponse startWebSign(final long memberId, final WebsignRequest request) {
+
+    val existing = signedMemberRepository.findBySsn(request.getSsn());
+
+    if (existing.isPresent()) {
+      throw new MemberHasExistingInsuranceException();
+    }
+
+    if (productApi.hasProductToSign(memberId)) {
+      val result = bankidService.startSign(memberId, request.getSsn(), "SomeMessage", request.getIpAddress());
```
Comment (meletios) [inline member-service/src/main/java/com/hedvig/memberservice/services/SigningService.java]: Should SomeMessage be replaced?

File: member-service/src/main/java/com/hedvig/memberservice/services/SigningService.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+  public void scheduleCollectJob(OrderResponse result) {
+    try {
+      val jobName = result.getOrderRef();
+      val jobDetail = newJob().withIdentity(jobName, "bankid.collect").ofType(
+          BankIdCollector.class).build();
+
+      val trigger = newTrigger()
+          .forJob(jobName, "bankid.collect")
+          .withSchedule(simpleSchedule().withIntervalInSeconds(1).withRepeatCount(900))
+          //.startAt(Date.from(Instant.now().plusSeconds(1)))
```
Comment (meletios) [inline member-service/src/main/java/com/hedvig/memberservice/services/SigningService.java]: remove comments

File: member-service/src/main/java/com/hedvig/memberservice/services/SigningService.java [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+            })
+        .orElseGet(
+            () -> {
+              log.error("Could not find SignSession with orderReference: ", orderReference);
+              return false;
+            });
+  }
+
+  public Optional<SignSession> getSignStatus(final long memberId, @NonNull final String orderRef) {
+    return signSessionRepository.findByOrderReference(orderRef);
```
Comment (meletios) [inline member-service/src/main/java/com/hedvig/memberservice/services/SigningService.java]: memberId is not used

File: member-service/src/main/java/com/hedvig/memberservice/web/v2/dto/WebsignRequest.java [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+import javax.validation.constraints.Email;
+import javax.validation.constraints.NotBlank;
+import lombok.Value;
+
+@Value
+public class WebsignRequest {
+
+  @Email
+  final String email;
```
Comment (hedvigoscar) [inline member-service/src/main/java/com/hedvig/memberservice/web/v2/dto/WebsignRequest.java]: `@Value` implies `final`, right? 🙂

**Pull request merged**